app/controllers/api/v2/registration_controller.rb

```diff
@@ -20,6 +20,7 @@ class RegistrationController < V2::BaseController
       param :hostgroup_id, :number, desc: N_("ID of the Host group to register the host in.")
       param :operatingsystem_id, :number, desc: N_("ID of the Operating System to register the host in.")
       param :setup_insights, :bool, desc: N_("Set 'host_registration_insights' parameter for the host. If it is set to true, insights client will be installed and registered on Red Hat family operating systems.")
+      param :setup_remote_execution, :bool, desc: N_("Set 'host_registration_remote_execution' parameter for the host. If it is set to true, SSH keys will be installed on the host.")
```

Comment:
I'm also wondering if REX is part of core or not. If it's not, this code belongs in the plugin repository.

Reply:
It's not a part of the core, and you are right that this should be done from plugin (`rex` or `insights`), but right now there is no easy way how we can hook from plugin into the controller action, like here:
https://github.com/theforeman/foreman/pull/8104/files#diff-216837bf690d5ab3bf5bca58373dd0b5969bb119fa157440382a056ce3ca0058R73-R80

cc @ares we talked about it,there isn't better way how to do it, right?

Reply:
It wouldn't be all that hard to place an extension point there, assuming you don't need to enforce specific ordering when extending from plugins.

Reply:
> It wouldn't be all that hard to place an extension point there

Can you point me to some docs or code, didn't know that we can make extension points inside of controllers.

Reply:
It is the same pattern as everywhere else. You define a method on the controller (say `host_registration_extension`), which will be a no-op by default and call it with params in place where you currently call `host_setup_insights` and `host_setup_remote_execution`. Plugins can then implement a module which overrides this method and include/prepend this module to the controller.

Reply:
I will have this:
```
# Foreman core
def controller_action
  ...
  extend_here
end

# Remote execution plugin
class KlassExtension
  def extend_here
    my_stuff1
  end
end

# Insights plugin
class KlassExtension
  def extend_here
    my_stuff2
  end
end
```

Won't there be an issue with overriding the method from each plugin? I mean, I need to call `my_stuff1` AND `my_stuff2`, but with code like this it will run only `my_stuff2`. Or no?

Reply:
If you wrote it like that, then it would indeed behave like that.

I was thinking about something along the lines of this
```ruby
# Foreman core
class ExtensibleController
  def extensible_action
    # ...
    extend_here
    # ...
  end

  def extend_here; end
end

# Remote execution plugin
module Rex::KlassExtension
  def extend_here
    super
    my_stuff1
  end
end

# in lib/rex/engine.rb
ExtensibleController.prepend Rex::KlassExtension

# Insights plugin
module Insights::KlassExtension
  def extend_here
    super
    my_stuff2
  end
end

# in lib/insights/engine.rb
ExtensibleController.prepend Insights::KlassExtension
```

Reply:
Just realized that there is another reason why we want to do this in the core:
Users may want to install & register insights on the hosts, but they don't use Insights plugin.
Same goes for the SSH keys - clients may want to deploy SSH keys on the host and they don't have to use rex plugin at all.

Reply:
@stejskalleos is correct, both insights and REX snippets work without any plugin and are today used in provisioning templates. REX plugins extends the functionality to provide SSH keys from proxy, however the snippet also works with pure Foreman, when user just set the key as a host parameter. Insights plugin provides very different functionality, this is a generic provisioning/registration feature to deploy the insights client.